delivery_carrier_preference/models/delivery_carrier_preference.py

```diff
@@ -0,0 +1,157 @@
+                        "must define a Delivery carrier."
+                    )
+                )
+        partner_pref_cnt = self.search_count(
+            [("preference", "=", "partner"), ("company_id", "=", self.env.company.id)]
+        )
+        if partner_pref_cnt > 1:
+            raise ValidationError(
+                _(
+                    "Only one Preferred Shipping Method can be set with "
+                    "'Partner carrier' preference."
+                )
+            )
+
+    @api.constrains("max_weight")
+    def _check_max_weight(self):
+        for pref in self:
+            if (
+                float_compare(
+                    pref.max_weight,
+                    0,
+                    precision_rounding=pref.max_weight_uom_id.rounding,
+                )
+                < 0
+            ):
+                raise ValidationError(
+                    _("Max weight must have a positive or null value.")
+                )
+
+    @api.onchange("preference")
+    def onchange_preference(self):
+        self.ensure_one()
+        if self.preference == "partner" and self.carrier_id:
+            self.carrier_id = False
+
+    @api.depends("preference", "carrier_id", "max_weight")
+    def _compute_name(self):
+        pref_descr = {
+            k: v for k, v in self._fields["preference"]._description_selection(self.env)
+        }
+        for pref in self:
+            name = pref_descr.get(pref.preference)
+            if pref.carrier_id:
+                name = _("%s: %s") % (name, pref.carrier_id.name)
+            if pref.max_weight:
+                name = _("%s (Max weight %s %s)") % (
+                    name,
+                    pref.max_weight,
+                    pref.max_weight_uom_id.display_name,
+                )
+            pref.name = name
+
+    def _default_max_weight_uom_id(self):
+        return self.env[
+            "product.template"
+        ]._get_weight_uom_id_from_ir_config_parameter()
+
+    def _compute_max_weight_uom_id(self):
+        for pref in self:
+            pref.max_weight_uom_id = self._default_max_weight_uom_id().id
+
+    @api.model
+    def get_preferred_carriers(self, partner, weight, company, picking=None):
+        # TODO Check possible conflicting settings between doc company and
+        #  user preference defined on another company?
+        company_carriers = self.env["delivery.carrier"].search(
+            ["|", ("company_id", "=", False), ("company_id", "=", company.id)]
+        )
+        carrier_preferences = self.search(
+            [
+                "&",
+                "|",
+                ("max_weight", ">=", weight),
+                ("max_weight", "=", 0.0),
+                "|",
+                ("carrier_id", "in", company_carriers.ids),
+                ("carrier_id", "=", False),
+            ]
+        )
+        carriers_ids = list()
+        for cp in carrier_preferences:
+            if (
+                picking is not None
```

Comment:
As you implemented, we cannot define rules at sales which set the default carrier when not manually chosen, and rules at release of operation that remap if weight < max_weight.
Either we should have 2 set of rules, or being able to flag what is applied at sales and only at sales. E.g. weight rule should not be applied at sales. Customer preferred carrier should not be applied at picking.